pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -165,46 +190,305 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+			if err != nil {
+				catchError(err)
+				cancel()
+				return
+			}
+			if ok {
+				for _, constraint := range constraints {
+					// constraint.TopologyKey is already guaranteed to be present
+					pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+					nodeTopologyMaps.addTopologyPair(pair, existingPod)
+				}
+			}
+		}
+		// If needed, append "headless" topology pair
+		// For example, on node-x, there is no pod matching spread constraints
+		// but node-x should be also considered as a match (with match number 0)
+		// i.e. <node: node-x>: {}
+		for _, constraint := range constraints {
+			// constraint.TopologyKey is already guaranteed to be present
+			pair := topologyPair{
+				key:   constraint.TopologyKey,
+				value: node.Labels[constraint.TopologyKey],
+			}
+			// addHeadlessTopologyPair is a non-op if other pods match this pair
+			nodeTopologyMaps.addHeadlessTopologyPair(pair)
+		}
+
+		appendTopologyPairsMaps(nodeTopologyMaps)
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+
+	// calculate min match for each topology pair
+	topologyPairsPodSpreadMap.minMatchMap = make(map[string]int32, len(constraints))
+	for _, constraint := range constraints {
+		topologyPairsPodSpreadMap.minMatchMap[constraint.TopologyKey] = MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
+		// TODO(Huang-Wei): short circuit to skip non-visited <topologyKey: any value> pairs
+		// if we see 0 as min match of the topologyKey
+		if l := int32(len(podSet)); l < topologyPairsPodSpreadMap.minMatchMap[pair.key] {
+			topologyPairsPodSpreadMap.minMatchMap[pair.key] = l
+		}
+	}
+	return topologyPairsPodSpreadMap
+}
+
+func getHardTopologySpreadConstraints(pod *v1.Pod) (constraints []v1.TopologySpreadConstraint) {
+	if pod != nil {
+		for _, constraint := range pod.Spec.TopologySpreadConstraints {
+			if constraint.WhenUnsatisfiable == v1.DoNotSchedule {
+				constraints = append(constraints, constraint)
+			}
+		}
+	}
+	return
+}
+
+func podMatchesAllSpreadConstraints(pod *v1.Pod, ns string, constraints []v1.TopologySpreadConstraint) (bool, error) {
+	if len(constraints) == 0 || pod.Namespace != ns {
+		return false, nil
+	}
+	return podLabelsMatchesSpreadConstraints(pod.Labels, constraints)
+}
+
+// some corner cases:
+// 1. podLabels = nil, constraint.LabelSelector = nil => returns false
+// 2. podLabels = nil => returns false
+// 3. constraint.LabelSelector = nil => returns false
+func podLabelsMatchesSpreadConstraints(podLabels map[string]string, constraints []v1.TopologySpreadConstraint) (bool, error) {
+	if len(constraints) == 0 {
+		return false, nil
+	}
+	for _, constraint := range constraints {
+		selector, err := metav1.LabelSelectorAsSelector(constraint.LabelSelector)
+		if err != nil {
+			return false, err
+		}
+		if !selector.Matches(labels.Set(podLabels)) {
```

Comment:
It seems the result of labels.Set(podLabels) can be stored in variable outside the loop since it is invariant across iterations

Reply:
I will update in the original PR which first introduced this change, and rebase here.